### PR TITLE
feat: Mermaid renderer for hugrs

### LINKED ISSUE #696: feat: automatic mermaid diagram generation
Mermaid diagrams are syntacticly simple and use indentation to follow the same hierarchical structure that HUGR has. The walker should be a perfect fit. Could replace the dot-string code base, or share the formatting code for names/ports etc.

Blocked by #626

### PR DESCRIPTION
**Pull request by aborgna-q**
Adds a `HugrView::mermaid_string` which produces things like
```mermaid
graph LR
    subgraph 0 ["(0) DFG"]
        direction LR
        1["(1) Input"]
        1--0:0-->3
        1--1:1-->3
        2["(2) Output"]
        3["(3) test.quantum.CX"]
        3--0:1-->4
        3--1:0-->4
        3-.2:2.-4
        4["(4) test.quantum.CX"]
        4--0:0-->2
        4--1:1-->2
    end
```
Note that edges in mermaid are unordered, so I had to add the port indices explicitly.

The new code in `src/hugr/views/render.rs` is just moved from `src/hugr/views.rs`.

Closes #696 

Requires https://github.com/CQCL/portgraph/pull/125

### CODE REVIEW

**Comment by aborgna-q** [comment]
Here's a render for a CFG:
```mermaid
graph LR
    subgraph 0 ["(0) CFG"]
        direction LR
        subgraph 2 ["(2) DataflowBlock"]
            direction LR
            3["(3) Input"]
            3--0:0-->5
            4["(4) Output"]
            5["(5) MakeTuple"]
            5--0:0-->6
            6["(6) Tag"]
            6--0:0-->4
        end
        2--0:0-->7
        2--1:0-->1
        1["(1) ExitBlock"]
        subgraph 7 ["(7) DataflowBlock"]
            direction LR
            8["(8) Input"]
            8--0:1-->9
            9["(9) Output"]
            10["(10) const:sum:{tag:0, val:const:seq:{}}"]
            10--0:0-->11
            11["(11) LoadConstant"]
            11--0:0-->9
        end
        7--0:0-->1
    end
```

I also tried adding the type to the edge labels. It's a bit more noisy, but probably more useful too.
```mermaid
graph LR
    subgraph 0 ["(0) CFG"]
        direction LR
        subgraph 2 ["(2) DataflowBlock"]
            direction LR
            3["(3) Input"]
            3--"0:0<br>usize([])"-->5
            4["(4) Output"]
            5["(5) MakeTuple"]
            5--"0:0<br>Tuple([usize([])])"-->6
            6["(6) Tag"]
            6--"0:0<br>Sum([Tuple([usize([])]), Tuple([usize([])])])"-->4
        end
        2--"0:0"-->7
        2--"1:0"-->1
        1["(1) ExitBlock"]
        subgraph 7 ["(7) DataflowBlock"]
            direction LR
            8["(8) Input"]
            8--"0:1<br>usize([])"-->9
            9["(9) Output"]
            10["(10) const:sum:{tag:0, val:const:seq:{}}"]
            10--"0:0<br>Sum(UnitSum(1))"-->11
            11["(11) LoadConstant"]
            11--"0:0<br>Sum(UnitSum(1))"-->9
        end
        7--"0:0"-->1
    end
```

And this is now the first example, with type info:
```mermaid
graph LR
    subgraph 0 ["(0) DFG"]
        direction LR
        1["(1) Input"]
        1--"0:0<br>qubit([])"-->3
        1--"1:1<br>qubit([])"-->3
        2["(2) Output"]
        3["(3) test.quantum.CX"]
        3--"0:1<br>qubit([])"-->4
        3--"1:0<br>qubit([])"-->4
        3-."2:2".-4
        4["(4) test.quantum.CX"]
        4--"0:0<br>qubit([])"-->2
        4--"1:1<br>qubit([])"-->2
    end
```

File: src/hugr/views/render.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+        let port_kind = src_optype.port_kind(src_offset).unwrap();
+
+        let style = if hugr.hierarchy.parent(src_node) != hugr.hierarchy.parent(tgt_node) {
+            EdgeStyle::Solid
+        } else if port_kind == EdgeKind::StateOrder {
+            EdgeStyle::Dotted
+        } else {
+            EdgeStyle::Solid
+        };
```
Comment (ss2165) [inline src/hugr/views/render.rs:70]: I'm not sure what the case distinctions intended here is, maybe add a comment?

Comment (ss2165) [reply src/hugr/views/render.rs:70]: Also can we distinguish CFG edges? in the spec we use yellow colour

File: src/hugr/views/render.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+        if !show_port_offsets {
+            return style;
+        }
+        let label = match port_kind {
+            EdgeKind::StateOrder | EdgeKind::ControlFlow => {
+                format!("{}:{}", src_offset.index(), tgt_offset.index())
+            }
+            EdgeKind::Static(ty) | EdgeKind::Value(ty) => {
+                format!("{}:{}\n{ty}", src_offset.index(), tgt_offset.index())
```
Comment (ss2165) [inline src/hugr/views/render.rs:80]: can types be optional?

File: src/hugr/views/snapshots/hugr__hugr__views__tests__mermaid_string.snap [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        1--"0:0<br>qubit([])"-->3
+        1--"1:1<br>qubit([])"-->3
+        2["(2) Output"]
+        3["(3) test.quantum.CX"]
+        3--"0:1<br>qubit([])"-->4
+        3--"1:0<br>qubit([])"-->4
+        3-."2:2".->4
+        4["(4) test.quantum.CX"]
+        4--"0:0<br>qubit([])"-->2
+        4--"1:1<br>qubit([])"-->2
```
Comment (ss2165) [inline src/hugr/views/snapshots/hugr__hugr__views__tests__mermaid_string.snap:19]: I don't think this tests all the possibles very well, maybe a more complicated test involving nested graphs, non-local static edges and a cfg?

**Comment by aborgna-q** [comment]
Here's the CFG rendered in the new test:

```mermaid
graph LR
    subgraph 0 ["(0) CFG"]
        direction LR
        subgraph 2 ["(2) DataflowBlock"]
            direction LR
            3["(3) Input"]
            3--"0:0<br>usize"-->5
            4["(4) Output"]
            5["(5) MakeTuple"]
            5--"0:0<br>Tuple([usize])"-->6
            6["(6) Tag"]
            6--"0:0<br>Sum([Tuple([usize]), Tuple([usize])])"-->4
        end
        2-."0:0".->7
        2-."1:0".->1
        1["(1) ExitBlock"]
        subgraph 7 ["(7) DataflowBlock"]
            direction LR
            8["(8) Input"]
            8--"0:1<br>usize"-->9
            9["(9) Output"]
            10["(10) const:sum:{tag:0, val:const:seq:{}}"]
            10--"0:0<br>Sum(UnitSum(1))"-->11
            11["(11) LoadConstant"]
            11--"0:0<br>Sum(UnitSum(1))"-->9
        end
        7-."0:0".->1
    end
```

File: src/hugr/views/render.rs [hunk 4]
```diff
@@ -40,45 +40,45 @@ pub fn port_style<H: HugrView>(h: &H) -> Box<dyn FnMut(PortIndex) -> PortStyle +
...
-        } else if port_kind == EdgeKind::StateOrder {
-            EdgeStyle::Dotted
-        } else {
-            EdgeStyle::Solid
+        // StateOrder edges: Dotted line.
+        // Control flow edges: Solid line.
+        // Static and Value edges: Solid line with label.
+        let style = match port_kind {
+            EdgeKind::StateOrder => EdgeStyle::Dotted,
+            EdgeKind::ControlFlow => EdgeStyle::Dashed,
```
Comment (ss2165) [inline src/hugr/views/render.rs:77]: comment says solid?

**Review by ss2165 — APPROVED** [review]

**Pull request merged**